comms/src/connection_manager/peer_connection.rs

```diff
@@ -222,6 +226,14 @@ impl PeerConnection {
         builder.connect(framed).await
     }
 
+    /// Creates a new RpcClientPool that can be shared between tasks. The client pool will lazily establish up to
+    /// `max_sessions` sessions and provides client session that is least used.
+    #[cfg(feature = "rpc")]
+    pub fn create_rpc_client_pool<T>(&self, max_sessions: usize) -> RpcClientPool<T>
```

Comment:
I would have thought that calls to `pub async fn connect_rpc` and `pub async fn connect_rpc_using_builder` would call `pub fn create_rpc_client_pool` if a pool does not exist end then return connections from the pool only? Is that not the purpose of the pool?

Reply:
I understand the thinking but that is not easy (or I'd argue desirable) to do given that all clients have a different type.

Remember a pool session is per peer _per protocol_.

A pool of sessions has to be typed to an rpc protocol   i.e. "RpcClientPool<WalletBaseNodeClient>". So storing a generic list containing different types requires `RpcClientPool<Box<dyn RpcPoolClient>>` which actually requires setting up a Boxed type. So basically,  making this generic is not straight forward - I also don't think it's a particularly good solution, because:

1. RPC is designed to for single concerns - one concern needs one kind of rpc service. Services are small and focused. You shouldn't need to open more than one/two sessions to a particular protocol i.e. you should design your software to not need more than this (though in reality you have quite a margin for abuse as we've seen). 
2. it is MUCH easier for the caller to get a pool and manage it, then to make a boxed dynamic list/hashmap.
3. Only the wallet needs a pool currently.
4. It has a memory and perf cost which is simply not necessary for most components to pay
5. Other minor concerns which add undue code complexity - When can we close the pooled sessions? Do we assume that we are never done? Do we add another call(s) to manage this pool state for everyone?

So, the wallet will get a pool, and then use it to get the connections it needs.

Reply:
Understood, thanks, and agreed. The caller/client process can/should manage the sessions.